notebooks/platform/xarray/lib/io/pybgen_backend.py

```diff
@@ -0,0 +1,91 @@
+import dask.array as da
+import numpy as np
+from pybgen import PyBGEN
+from pybgen.parallel import ParallelPyBGEN
+import xarray as xr
+
+from .. import core
+from ..typing import PathType
+from ..compat import Requirement
+from ..dispatch import ClassBackend, register_backend
+from .core import BGEN_DOMAIN
+
+
+def _array_name(f, path):
+    return f.__qualname__ + ':' + str(path)
+
+
+class BgenReader(object):
+
+    def __init__(self, path, dtype=np.float32):
+        self.path = str(path) # pybgen needs string paths
+
+        # Use ParallelPyBGEN only to get all the variant seek positions from the BGEN index.
+        # No parallel IO happens here.
+        with ParallelPyBGEN(self.path, probs_only=False) as bgen:
+            bgen._get_all_seeks()
+            self._seeks = bgen._seeks
+            n_variants = bgen.nb_variants
+            n_samples = bgen.nb_samples
+
+            self.shape = (n_variants, n_samples)
+            self.dtype = dtype
+            self.ndim = 2
+
+            self.sample_id = bgen.samples
+            # This may need chunking for large numbers of variants
+            variants = list(bgen.iter_variant_info())
```

Comment:
Is it possible to get the number of alleles for each variant here?  This wasn't something we had to worry about with the PLINK reader since it only supports biallelic variants.  I think we should decide if we want to support multi-allelic bgen data from the start (in which case we have to rethink the `a1`/`a2` fields below), or if we want to only allow biallelic variants and possibly check the allele counts here to throw an error otherwise.

Reply:
Good question. My reading of PyBGEN is that it only support biallelic variants, and will fail if the BGEN file does not conform to this. So we are currently limited to biallelic, but should document it somewhere (any suggestions where?).

Reply:
Ahh I see.  I like what Hail did for this in putting limitations like that into the doc strings on the associated methods.  So io/core.py#read_ben would be my vote since that's the documentation users should see on autocomplete and in a RTD site.

Reply:
Done